### PR TITLE
unigen; as_set and non-Symbols; as_relational and +/-oo

### LINKED ISSUE #22073: Interval with oo
Hi:
when use Interval with oo, can not get Lopen or Ropen interval.
```
Interval.Lopen(-oo, 5)
Out[1]: Interval(-oo, 5)
Interval.Ropen(5, +oo)
Out[2]: Interval(5, oo)
```
how to solve this？
Thanks

### LINKED ISSUE #22175: Incorrect results with Piecewise expression MatrixSymbol substitutions.
This follows on from #14933. Feel free to delete if it's preferred that that issue be reopened. 

I'm encountering strange results with the `master` branch version of Sympy:

```python
from sympy import Piecewise, Abs, Symbol, MatrixSymbol
x = Symbol("x")
fn = Piecewise((1.0, Abs(x) < 0.5), (3., (Abs(x - 2) <= 0.5) | (Abs(x- 1) < 0.5)), (1.0, True))
print(fn)
# Piecewise((1.0, Abs(x) < 0.5), (3.0, (x <= 2.5) & (x > 0.5)), (1.0, True))
mat_x = MatrixSymbol(  'mat_x', 1,1)
subfn = fn.xreplace({x:mat_x[0,0],})
print(subfn)
# 1.0
```
Against `sympy==1.8` this test produces the `TypeError: 'Dummy' object is not subscriptable` error discussed in #14933. 

Thanks

### PR DESCRIPTION
**Pull request by smichr**
#### Brief description of what is fixed or changed

fixes #22175 by introducing `unigen` which looks for the smallest symbol-like univariate generator for an expression.

fixes #22073 by returning a union to show inclusion of infinity, Relationals are unbounded and hold in extended real whereas `Interval(-oo,oo) == Reals`
```python
>>> (x>9).as_set()
Union({oo}, Interval.open(9, oo))
>>> _.as_relational(x)
x > 9
```
To maintain economy when dealing with such sets being conveted to Relationals, special tests are done to return something similar to the starting Relational. So instead of `Eq(x, oo) | (x > 9) & (x < oo)` for the above, `x > 9` is returned.

With this in place, the corner-case introduction of `=` to relationships expressed without equality in Piecewise conditions has been fixed.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/sets/sets.py [hunk 1]
```diff
@@ -1192,6 +1192,9 @@ def __new__(cls, *args, **kwargs):
         # flatten inputs to merge intersections and iterables
         args = _sympify(args)
 
+        # replace Interval(-oo, oo) with S.Reals
```
Comment (smichr) [inline sympy/sets/sets.py:1195]: Perhaps better to just do this in the `Union.as_relational` routine?
